### PR TITLE
editoast: refactoring STDCM handler

### LINKED ISSUE #9476: editoast: refactoring STDCM handler
### Description and goal

Refactored STDCM handler and addressed comments in this [PR](https://github.com/OpenRailAssociation/osrd/pull/9327).

### Acceptance criteria

Fix all comments.

### PR DESCRIPTION
**Pull request by hamz2a**
closes https://github.com/OpenRailAssociation/osrd/issues/9476

### CODE REVIEW

File: front/src/applications/stdcm/hooks/useStdcm.ts
Comment (SharglutDev) [inline front/src/applications/stdcm/hooks/useStdcm.ts]: This might conflict with @achrafmohye PR : https://github.com/OpenRailAssociation/osrd/pull/9419

Comment (hamz2a) [reply front/src/applications/stdcm/hooks/useStdcm.ts]: Ok, I’ll wait for this PR to be merged before.

Comment (hamz2a) [reply front/src/applications/stdcm/hooks/useStdcm.ts]: I rebased, so there's no more conflict.

**Comment by eckter** [comment]
Before checking the code itself: do we want to merge it during the feature freeze? Are there bug fixes in there, or is it just moving code around?

**Comment by hamz2a** [comment]
> Before checking the code itself: do we want to merge it during the feature freeze? Are there bug fixes in there, or is it just moving code around?

This is purely a refactoring with no functional changes.
Are there any risks you see that would block merging it during the freeze?

File: editoast/src/core/stdcm.rs
Comment (woshilapin) [inline editoast/src/core/stdcm.rs]: Follow up of https://github.com/OpenRailAssociation/osrd/pull/9434#discussion_r1816331474. I still believe that both should be named `STDCMResponse`, one will be named `core::STDCMResponse` and the other one will be `views::timetable::stdcm::STDCMResponse`.

**Review by woshilapin — APPROVED** [review]
Thanks for the addressed comments from #9434.

**Comment by eckter** [comment]
> This is purely a refactoring with no functional changes.
> Are there any risks you see that would block merging it during the freeze?

The same as any kind of change, bugs. I believe refactorings are even more "unstable" than new features as they affect what's already there and not just new things.

This specific part of the code that translates front inputs into core payloads is especially sensitive. It's poorly tested and it's difficult to track problems there, and we've had *issues* recently. At a time where we're working on stabilizing stdcm, it would be painful to introduce instability.

Let's talk about this on matrix, it's not about this PR specifically.

File: editoast/openapi.yaml
Comment (leovalais) [inline editoast/openapi.yaml:2638]: Since this is a refactoring I'd argue that the public API shouldn't change. Is there a good reason this variant was removed?

File: editoast/src/core/simulation.rs
Comment (leovalais) [inline editoast/src/core/simulation.rs]: Why the `Box<Self>` instead of just `Self`?

File: editoast/src/core/simulation.rs
Comment (leovalais) [inline editoast/src/core/simulation.rs]: Are we 100% sure that Core doesn't send an empty list? If so I'd suggest making the message a bit more explicit

```suggestion
                Ok(*provisional.times.last().expect("core error: empty simulation result"))
```

If not, probably best to return an `Option<u64>` and let the caller deal with the situation.

File: editoast/src/core/stdcm.rs
Comment (leovalais) [reply editoast/src/core/stdcm.rs]: String agree. I'd also suggest to rename this `StdcmResponse` while doing so to respect the case (I know it's an acronym, but look at `Cow<>` for instance).

File: editoast/src/core/stdcm.rs
Comment (leovalais) [inline editoast/src/core/stdcm.rs:133]: Why removing this?

File: editoast/src/models/work_schedules.rs
Comment (leovalais) [inline editoast/src/models/work_schedules.rs]: I'd suggest merging these two functions into a single `fn as_core_work_schedule(&self, start_time, latest_simulation_end) -> Option<core::stdcm::WorkSchedule>`. A function which cannot be named better that `do_this_and_do_that` is often an indicator of something sketchy.

Since you're not taking ownership of the `WorkSchedule`, the caller can still access its `id` afterwards.

Wdyt?

File: editoast/src/models/work_schedules.rs
Comment (leovalais) [inline editoast/src/models/work_schedules.rs]: big nit, feel free to ignore:

```suggestion
fn elapsed_time_since_ms(time: &NaiveDateTime, since: &DateTime<Utc>) -> u64 {
```

Also can't we have `time: &Datetime<Utc>` for consistency?

File: editoast/src/views/timetable.rs
Comment (leovalais) [inline editoast/src/views/timetable.rs]: Wdyt about moving all that into the `stcdm` module?

File: editoast/src/views/timetable/path_not_found_handler.rs
Comment (leovalais) [inline editoast/src/views/timetable/path_not_found_handler.rs:1]: I'm a bit conflicted about the name of the struct: `PathNotFoundHandler` strikes me like a handler for an `Err` case, where it's actually what we do when the simulation "successfully failed". I could be wired wrong though x)

Wdty about?

* Renaming `PathNotFoundHandler` into `SimulationFailureHandler`
* Renaming the module `failure_handler.rs`

@woshilapin

File: editoast/src/views/timetable/path_not_found_handler.rs
Comment (leovalais) [inline editoast/src/views/timetable/path_not_found_handler.rs]: This structure is important (it drives this whole PR after all). It should be documented properly:

1. In which case it's useful
2. What it does
3. What a "virtual train" is

Also please restrict the visibility to the most restrictive scope `pub(super)` or something else.

File: editoast/src/views/timetable/path_not_found_handler.rs
Comment (leovalais) [inline editoast/src/views/timetable/path_not_found_handler.rs]:
```suggestion
    pub(super?) async fn compute_conflicts(self) -> Result<STDCMResponse> {
```

File: editoast/src/views/timetable/path_not_found_handler.rs
Comment (leovalais) [inline editoast/src/views/timetable/path_not_found_handler.rs]:
```suggestion
        let train_schedules = [
            self.train_schedules,
            vec![self.virtual_train_schedule],
        ]
        .concat();

        // Combine the original simulations with the virtual train's simulation results.
        let simulations = [
            self.simulations,
            vec![self.virtual_train_sim_result],
        ]
        .concat();
```

Comment (leovalais) [reply editoast/src/views/timetable/path_not_found_handler.rs]: Also `self.train_schedules.push(...)`?

File: editoast/src/views/timetable/path_not_found_handler.rs
Comment (leovalais) [inline editoast/src/views/timetable/path_not_found_handler.rs]:
```suggestion
            expected_version: self.infra_version,
```

File: editoast/src/views/timetable/path_not_found_handler.rs
Comment (leovalais) [reply editoast/src/views/timetable/path_not_found_handler.rs:1]: Also what about renaming changing "handler" to "fallback" as the term handler also has a meaning for `axum`, and we're in `views`?

File: editoast/src/views/timetable/stdcm.rs
Comment (leovalais) [inline editoast/src/views/timetable/stdcm.rs]: `pub(in `

File: editoast/src/views/timetable/stdcm.rs
Comment (leovalais) [inline editoast/src/views/timetable/stdcm.rs]:
```suggestion
    let simulation_run_time = match virtual_train_sim_result.simulation_run_time() {
```

The syntax `let Ok(..) = .. else ` covers this precise pattern ;)

File: editoast/src/core/simulation.rs
Comment (leovalais) [reply editoast/src/core/simulation.rs]: Also
```suggestion
    pub fn simulation_run_time(&self) -> Result<u64, Box<Self>> {
```

File: editoast/src/views/timetable/stdcm.rs
Comment (leovalais) [inline editoast/src/views/timetable/stdcm.rs]: I'm not sure this documentation is useful as it doesn't bring anything we can't read in the return type.

File: editoast/src/views/timetable/stdcm.rs
Comment (leovalais) [inline editoast/src/views/timetable/stdcm.rs]: Shouldn't we have a `VirtualTrainRun` structure aggregating the information of the virtual train? That would also reduce the number of fields of `PathNotFoundHandler`.

File: editoast/src/views/timetable/stdcm_request_payload.rs
Comment (leovalais) [inline editoast/src/views/timetable/stdcm_request_payload.rs:1]: Likewise about visibility

File: editoast/src/views/timetable/stdcm_request_payload.rs
Comment (leovalais) [inline editoast/src/views/timetable/stdcm_request_payload.rs]: A bit more bike shed on the naming: we're in `views`, I'd say we don't need the `Payload` suffix to understand this struct refers to the content of the http request of the STDCM endpoint. 

Additionally, I'd rather have this module to be included in the `stdcm` module rather than `timetable` directly (see previous comment). In this case I'd argue that `Request` is a good enough name: both `stdcm/request.rs` and `stdcm::Request` are explanatory enough.

Just like for the `core` definitions and the `Core` prefix, `views::timetable::stcdm` is already providing a lot of context.

Wdyt?

File: editoast/src/views/timetable/stdcm_request_payload.rs
Comment (leovalais) [inline editoast/src/views/timetable/stdcm_request_payload.rs]:
```suggestion
    }

```

**Review by leovalais — CHANGES_REQUESTED** [review]
Thanks for all this work, it looks much more organized now! I left a bunch of comments which are mostly about:

* module organization
* visibility being more permissive than required (I didn't add a comment in all places)
* naming

It would be nice to have some module-level documentation in `stdcm.rs` describing the whole workflow still, even if it's much clearer now thanks to structure.

I'm not as worried about bugs as @eckter is: we already released a tested version and we're shooting demo videos very soon. I'd say that if a bug arises we're better off fixing it in the refactored code directly since this won't impact the PIP. I'm fine with waiting until the end of the week before merging in case a critical bug happens preventing us from shooting the videos, but there's no point in waiting after that point imo. Wdyt?

**Review by SharglutDev — APPROVED** [review]
Lgtm for front part (not tested, we were'nt handling this status anymore)

File: editoast/openapi.yaml
Comment (hamz2a) [reply editoast/openapi.yaml:2638]: This variant is in Core's response, but I created a new struct for `StdcmResponse` in views, where this variant isn’t included, so we’re replacing it with `conflicts`.

File: editoast/src/core/simulation.rs
Comment (hamz2a) [reply editoast/src/core/simulation.rs]: I updated the error message.
We need the result to forward it within this part of the code.
```rust
    let simulation_run_time = match virtual_train_sim_result.clone().simulation_run_time() {
        Ok(value) => value,
        Err(error) => {
            return Ok(Json(StdcmResponse::PreprocessingSimulationError {
                error: *error,
            }))
        }
    };
```

File: editoast/src/core/stdcm.rs
Comment (hamz2a) [reply editoast/src/core/stdcm.rs:133]: Core isn't returning conflicts for the STDCM request.

File: editoast/src/models/work_schedules.rs
Comment (hamz2a) [reply editoast/src/models/work_schedules.rs]: Done.

File: editoast/src/models/work_schedules.rs
Comment (hamz2a) [reply editoast/src/models/work_schedules.rs]: Done.

File: editoast/src/core/simulation.rs
Comment (hamz2a) [reply editoast/src/core/simulation.rs]: I had this warning:
```bash
pub fn simulation_run_time(self) -> Result<u64, Self> {
    |                                           ^^^^^^^^^^^^^^^^^
    |
    = help: try reducing the size of `core::simulation::SimulationResponse`, for example by boxing large elements or replacing it with `Box<core::simulation::SimulationResponse>`
```

Comment (leovalais) [reply editoast/src/core/simulation.rs]: Oh alright then. Maybe add a comment explaining the presence of the Box since this is not straightforward (to me at least).

File: editoast/src/core/simulation.rs
Comment (leovalais) [reply editoast/src/core/simulation.rs]: > If not, probably best to return an Option<u64> and let the caller deal with the situation.

I meant to return an `Ok(Some(...))` instead, that was unclear, sorry

File: editoast/src/views/timetable.rs
Comment (hamz2a) [reply editoast/src/views/timetable.rs]: Done.

File: editoast/src/views/timetable/path_not_found_handler.rs
Comment (hamz2a) [reply editoast/src/views/timetable/path_not_found_handler.rs:1]: I prefer to keep `handler`; I find this name clearer. Wdyt ?
Otherwise, I've changed the module and struct names.

File: editoast/src/views/timetable/path_not_found_handler.rs
Comment (hamz2a) [reply editoast/src/views/timetable/path_not_found_handler.rs]: Done.

File: editoast/src/views/timetable/path_not_found_handler.rs
Comment (hamz2a) [reply editoast/src/views/timetable/path_not_found_handler.rs]: To do `self.train_schedules.push(...)`, `self` needs to be mutable.
Therefore, I decided to clone certain parts of `self` instead of making the entire self mutable.

File: editoast/src/views/timetable/path_not_found_handler.rs
Comment (hamz2a) [reply editoast/src/views/timetable/path_not_found_handler.rs]: i got this error:
```bash
   |
93 |             expected_version: self.infra_version,
   |                               ------------------ value partially moved here
94 |             trains_requirements,
95 |             work_schedules: self.get_work_schedules_request(),
   |                             ^^^^ value borrowed here after partial move
   |
   = note: partial move occurs because `self.infra_version` has type `std::string::String`, which does not implement the `Copy` trait
 ```

File: editoast/src/views/timetable/stdcm.rs
Comment (hamz2a) [reply editoast/src/views/timetable/stdcm.rs]: Done.

File: editoast/src/views/timetable/stdcm_request_payload.rs
Comment (hamz2a) [reply editoast/src/views/timetable/stdcm_request_payload.rs]: Thanks for the great feedback! I agree with your points. I'll go ahead and remove the `Payload` suffix and move the module to `stdcm`. `Request` sounds perfect, and the context is clear.

Comment (hamz2a) [reply editoast/src/views/timetable/stdcm_request_payload.rs:71]: Thanks for the great feedback! I agree with your points. I'll go ahead and remove the `Payload` suffix and move the module to `stdcm`. `Request` sounds perfect, and the context is clear.

File: editoast/src/views/timetable/path_not_found_handler.rs
Comment (hamz2a) [reply editoast/src/views/timetable/path_not_found_handler.rs]: Done.

File: editoast/src/views/timetable/stdcm.rs
Comment (hamz2a) [reply editoast/src/views/timetable/stdcm.rs]: Done.

File: editoast/src/views/timetable/path_not_found_handler.rs
Comment (woshilapin) [reply editoast/src/views/timetable/path_not_found_handler.rs:1]: Agreed about it's more about simulation... but it's for the special case of path not found. Do we think we might have another handler for another simulation error? I'd say maybe, but if you think we can stick all in this one, I'm fine with `SimulationFailureHandler`.

As for "handler" vs "fallback", I'll stick with @hamz2a on this one for 2 reasons:
- I indeed find "handler" clearer
- `axum` does have the notion of "handler" but it's not used anywhere in the code (almost, try a `grep -i handler editoast/` to see what I mean). And indeed, in `axum`, `Handler` is a trait (but most likely, we never need it, not even need to import it), `handler` is a module (might use it in a `use`), and there is `#[debug_handler]` (which should never really stick in production code).

Comment (leovalais) [reply editoast/src/views/timetable/path_not_found_handler.rs:1]: Ok, let's keep `-Handler` then.

File: editoast/src/views/timetable/path_not_found_handler.rs
Comment (leovalais) [reply editoast/src/views/timetable/path_not_found_handler.rs]: The function takes ownership of `self` and never returns it. I'd say it's fine to make it `mut` since it's already clear that the `SimulationFailureHandler` won't live after the call. Wdyt?

File: editoast/src/views/timetable/path_not_found_handler.rs
Comment (leovalais) [reply editoast/src/views/timetable/path_not_found_handler.rs]: `self.get_work_schedules_request` doesn't move `self`, I guess you can swap the two fields in the struct initializer and avoid the clone

Comment (hamz2a) [reply editoast/src/views/timetable/path_not_found_handler.rs]: Thanks for the suggestion! Swapping the fields did the trick, and it worked perfectly without needing the clone. Great idea, and I appreciate the insight!

File: editoast/src/views/timetable/path_not_found_handler.rs
Comment (hamz2a) [reply editoast/src/views/timetable/path_not_found_handler.rs]: I tested using `mut`, but it caused excessive cloning, so I tried destructuring instead. Let me know what you think.

File: editoast/src/views/timetable/stdcm.rs
Comment (leovalais) [inline editoast/src/views/timetable/stdcm.rs]:
```suggestion
enum StdcmError {
```

File: editoast/src/core/simulation.rs
Comment (leovalais) [reply editoast/src/core/simulation.rs]: Up https://github.com/OpenRailAssociation/osrd/pull/9509/files#r1829667767

File: editoast/src/views/timetable/stdcm/failure_handler.rs
Comment (leovalais) [inline editoast/src/views/timetable/stdcm/failure_handler.rs]:
```suggestion
pub(super) struct SimulationFailureHandler {
```

File: editoast/src/views/timetable/stdcm/failure_handler.rs
Comment (leovalais) [inline editoast/src/views/timetable/stdcm/failure_handler.rs]:
```suggestion
        let Self {
            mut train_schedules,
            mut simulations,
            virtual_train_run:
                VirtualTrainRun {
                    train_schedule,
                    simulation,
                    pathfinding,
                },
            infra_id,
            infra_version,
            earliest_departure_time,
            latest_simulation_end,
            ..
        } = self;

```

You can destructure several levels at once.

File: editoast/src/views/timetable/path_not_found_handler.rs
Comment (leovalais) [reply editoast/src/views/timetable/path_not_found_handler.rs]: I'm not sure why this would cause any legitimate `clone()`, can you provide an example?

Destructuring is fine by me, but the workflow became weird with the `let work_schedules = self.get_work_schedules_request();` that has to occur before that.

I'd suggest either:

1. Try to use `mut self` and keep the call to `self.get_work_schedules_request` where it was before
2. Make `get_work_schedules_request` a function that takes all the `self` data as parameters and call it using the variables created by the destructuring.

File: editoast/src/views/timetable/path_not_found_handler.rs
Comment (leovalais) [reply editoast/src/views/timetable/path_not_found_handler.rs]: Cf. https://github.com/OpenRailAssociation/osrd/pull/9509#discussion_r1834761313

File: editoast/src/views/timetable/stdcm/failure_handler.rs
Comment (leovalais) [inline editoast/src/views/timetable/stdcm/failure_handler.rs]: `Option.map`?

File: editoast/src/views/timetable/stdcm/request.rs
Comment (leovalais) [inline editoast/src/views/timetable/stdcm/request.rs:1]: Up about visibility.

File: editoast/src/views/timetable/stdcm.rs
Comment (leovalais) [inline editoast/src/views/timetable/stdcm.rs:1]: Likewise about visibility in this file.

File: editoast/src/views/timetable/stdcm.rs
Comment (leovalais) [inline editoast/src/views/timetable/stdcm.rs]: Moving `train_requirements` and `speed_limit_tags` further down the field list may make the `clone()` unnecessary.

File: editoast/src/views/timetable/stdcm.rs
Comment (leovalais) [inline editoast/src/views/timetable/stdcm.rs]:
```suggestion
    async fn simulate(
```

**Review by leovalais — COMMENTED** [review]
Thanks for the changes!

File: editoast/src/views/timetable/stdcm.rs
Comment (hamz2a) [reply editoast/src/views/timetable/stdcm.rs]: Done.

File: editoast/src/views/timetable/stdcm/failure_handler.rs
Comment (hamz2a) [reply editoast/src/views/timetable/stdcm/failure_handler.rs]: Done.

File: editoast/src/views/timetable/stdcm/failure_handler.rs
Comment (hamz2a) [reply editoast/src/views/timetable/stdcm/failure_handler.rs]: Done.

File: editoast/src/views/timetable/stdcm.rs
Comment (hamz2a) [reply editoast/src/views/timetable/stdcm.rs]: Done.

File: editoast/src/views/timetable/stdcm/request.rs
Comment (hamz2a) [reply editoast/src/views/timetable/stdcm/request.rs:1]: Done.

File: editoast/src/views/timetable/stdcm.rs
Comment (hamz2a) [reply editoast/src/views/timetable/stdcm.rs:1]: Done.

File: editoast/src/views/timetable/stdcm/request.rs
Comment (hamz2a) [reply editoast/src/views/timetable/stdcm/request.rs:1]: Done.

File: editoast/src/views/timetable/stdcm/request.rs
Comment (eckter) [inline editoast/src/views/timetable/stdcm/request.rs:155]: That feels wrong, I'd expect this instead:
```suggestion
            .unwrap_or(self.get_earliest_step_tolerance_window())
```

---

@Erashin this function has just been moved here and you wrote the previous version, do you remember if there's a reason to add the simulation run time there?

File: editoast/src/views/timetable/stdcm/request.rs
Comment (eckter) [inline editoast/src/views/timetable/stdcm/request.rs:161]: Something is likely wrong there as well... We're supposed to return the simulation duration where the time spent between stops has been doubled. Here I believe stop duration is already included in `simulation_run_time`.

If a train leaves at 10:00, stops from 13:00 to 13:05 then arrives at 14:05. We spend 2h moving and 5 minutes stopped: we should return `2 * 2h + 5m = 4:05`. But in this version we may be returning `2 * 2h05 + 5m = 4:15`.

I realize that this isn't part of the PR and the code has just been moved, I'll test this on the side.

**Review by eckter — APPROVED** [review]
Much clearer, thanks :+1: 

I've noticed a couple weird things (that were already there in the previous version), I've commented them and I'll check further

File: editoast/src/views/timetable/stdcm/failure_handler.rs
Comment (hamz2a) [reply editoast/src/views/timetable/stdcm/failure_handler.rs]: Done.

**Review by leovalais — APPROVED** [review]
LGTM. Thanks for the changes!

File: editoast/src/views/timetable/stdcm/request.rs
Comment (Erashin) [reply editoast/src/views/timetable/stdcm/request.rs:155]: In the case where the departure is provided, that is indeed the case. However, if it's the end that is provided, it will always fail if the simulation run time is a lot higher than the tolerance window.. the best would be to differentiate both cases :
- if the departure is provided, max departure delay should be the tolerance window
- if it's the end, it should be as specified right now.

**Pull request merged**
